CMakeLists.txt

```diff
@@ -128,7 +130,21 @@ list(APPEND CMAKE_MODULE_PATH "${PROJECT_SOURCE_DIR}/cmake")
 if(WIN32)
   add_library(ONEMKL::SYCL::SYCL INTERFACE IMPORTED)
 else()
-  find_package(Compiler REQUIRED)
+  # Find necessary packages
+  string( TOLOWER "${ONEMKL_SYCL_IMPLEMENTATION}" ONEMKL_SYCL_IMPLEMENTATION)
+  if (ONEMKL_SYCL_IMPLEMENTATION STREQUAL "hipsycl")
```

Comment:
Do you think there is a way to detect compiler automatically? At this moment `CMAKE_CXX_COMPILER` is already defined so we could try to check what type of sycl implementation it is in case if user didn't provide `ONEMKL_SYCL_IMPLEMENTATION`.

Reply:
I am not sure how this could be done in a straightforward way. In the case of hipSYCL, the compiler used for the compilation is not necessary will be the compiler set in `CMAKE_CXX_COMPILER`. The `add_sycl_to_target` cmake command, will invoke the `syclcc-launcher` that will, in turn, invoke the compiler configured for hipSYCL during installation (eg clang). Therefore I don't see how this could be done, easily. Pinging @illuhad to fact check this.

Reply:
Sorry for the delay - it's indeed not easily possible to detect a SYCL implementation from the compiler in general. SYCL implementations have quite a lot of freedom in how they are designed, and e.g. in certain split compilation scenarios the `CMAKE_CXX_COMPILER` only plays a minor role.

The fundamental issue is that there's no global cmake infrastructure that works throughout the entire SYCL ecosystem, like a `find_package(SYCL)`. Or to be more specific, that DPC++ has not yet adopted `add_sycl_to_target()` which is used by ComputeCpp, hipSYCL and triSYCL due to the flexibility of this approach.

See here and the associated MR for a discussion of the SYCL ecosystem cmake infrastructure, and the difficulties of the assumption that SYCL compilation only requires `CMAKE_CXX_COMPILER` + special compiler flags:
https://gitlab.kitware.com/cmake/cmake/-/issues/21711

hipSYCL indeed does not necessarily use `CMAKE_CXX_COMPILER`. This actually has advantages for multi-language projects, where e.g. some part is OpenMP offload, another part SYCL, another CUDA etc.

Reply:
Hi @illuhad and @sbalint98 . Sorry for the delay! Let me provide one possible suggestion how we can potentially add auto detection for compiler via `execute_process`. I rely on `CMAKE_CXX_COMPILER` to detect vendor, but if hipSYCL doesn't provide this information I guess this option wouldn't work. But if you know any other ways how to get information about vendor from hipSYCL compiler from command line please share, we can try them as well.

Reply:
Thanks for the suggestion, @mkrainiuk. I'm not sure how well this can work. hipSYCL supports multiple compilation flows, in particular it can also act as a mere library for third-party compilers when targeting CPU exclusively. In this case, it can happen that the underlying compiler changes depending on which targets are selected. For example:

```
$ syclcc --hipsycl-targets="omp" --version
c++ (GCC) 11.1.0
Copyright (C) 2021 Free Software Foundation, Inc.
Dies ist freie Software; die Kopierbedingungen stehen in den Quellen. Es
gibt KEINE Garantie; auch nicht für MARKTGÄNGIGKEIT oder FÜR SPEZIELLE ZWECKE.
$ syclcc --hipsycl-targets="hip:gfx908" --version
clang version 11.1.0
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
```

cmake tends to not like it when the compiler suddenly changes depending on compiler arguments (e.g. this might also cause certain compiler specific flags to suddenly be not available anymore), which is why it is generally *not* considered good practice to set `CMAKE_CXX_COMPILER` to hipSYCL's `syclcc` even though it might work for some use cases. The better approach is to just use `add_sycl_to_target()` without changing `CMAKE_CXX_COMPILER`. As such, relying on `CMAKE_CXX_COMPILER` for implementation detection is probably not an ideal strategy.

The better approach in the long run in my opinion would probably be for the SYCL implementation ecosystem to come together and have a unified `FindSYCL.cmake` which checks for available SYCL implementations in the filesystem and then provides a unified interface to compile targets with SYCL. But we are still some time away from this, and while this is the case a cmake configuration flag to select the implementation as done here is probably the most straight-forward solution.

Reply:
Thanks @illuhad for detailed explanation! So it's more DPC++ compiler problem, because user has to source DPC++ compiler and define it as cmake compiler, one more variable looks like an overkill. What do you think about having autodetection mechanism for DPC++ compiler and if it's not DPC++ compiler use hipSYCL check? The proposed option is also nice to have to force compiler without check, but I'd prefer to make it optional for better user experience.

Reply:
@mkrainiuk yes, I think it could work to first try to automatically detect DPC++ from `CMAKE_CXX_COMPILER`, and if that fails, try `find_package(hipSYCL)`.

I think we should still add an optional option to force a specific compiler because I expect that the line between DPC++ and mainline clang will get blurry in the future as more components from DPC++ get upstreamed. In the future, there will probably be 
* people using regular clang, who want to run the SYCL path for DPC++
* and be people who happen to use clang (with SYCL support) as `CMAKE_CXX_COMPILER`, but still wish to use hipSYCL.

Reply:
Sure, I'm fine with new option for advanced users, but we need some default detection for easier way to start using this project.
One more question out of curiosity, how hipSYCL plans to support several compilation runs for host and kernels without special compiler? As far as I know standard compiler can't do host/kernels compilation from one source.

Reply:
> One more question out of curiosity, how hipSYCL plans to support several compilation runs for host and kernels without special compiler? As far as I know standard compiler can't do host/kernels compilation from one source.

Currently, as soon as non-CPU devices are targeted as well, hipSYCL forces compilation with clang and hipSYCL's custom clang plugin to perform kernel outlining and device compilation etc. Currently using third-party compilers is only possible when targeting CPU exclusively, although we may also enable this in the future using nvc++ or OpenMP targets.

Reply:
I rebased to develop and updated with the required logic